hash.c

```diff
@@ -2887,7 +2887,7 @@ rb_hash_key_str(VALUE key)
 static int
 hash_aset_str(st_data_t *key, st_data_t *val, struct update_arg *arg, int existing)
 {
-    if (!existing && !RB_OBJ_FROZEN(*key)) {
+    if (!existing && !FL_TEST(*key, RSTRING_FSTR)) {
```

Comment:
Bonus: if you could make sure `*key` can never be a non-string (I guess so), you can use `FL_TEST_RAW` instead to eliminate 2 more branches.

Reply:
Right, this function is called inside a `rb_obj_class(key) != rb_cString)` condition. I suppose that's enough?